File: LeetCode/简单/392. 判断子序列/Solution392.java
/*
url: https://leetcode.cn/problems/is-subsequence/description/
相关: LeetCode3132. 找出与数组相加的整数 II
标签: 双指针, 同向双指针, 子序列, 判断子序列
*/

// `LeetCode3132. 找出与数组相加的整数 II`用到此题
public class Solution392 {
    public boolean isSubsequence(String s, String t) {
        int i = 0, j = 0; // 用双指针来匹配
        while (i < s.length() && j < t.length()) {
            if (s.charAt(i) == t.charAt(j)) {
                i++;
                j++;
            }
            else {
                // 碰到相等就可以直接到下一个匹配位置，不存在选第一个不能匹配完而选后面的匹配位置能匹配上的可能
                j++;
            }
        }
        return i == s.length();
    }

    public static void main(String[] args) {
        Solution392 solu = new Solution392();
        System.out.println(solu.isSubsequence("abc", "ahbgdc"));
    }
}

File: LeetCode/简单/392. 判断子序列/solution392.cpp
/*
url: https://leetcode.cn/problems/is-subsequence/
标签: 【题单】滑动窗口与双指针
*/
#include <bits/stdc++.h>

using namespace std;

class Solution392 {
public:
    bool isSubsequence(string s, string t) { // 判断s是否为t的子序列
        int i = 0, j = 0;
        while (i < s.size() && j < t.size()) {
            if (s[i] == t[j]) {
                i++; j++;
            }
            else j++;
        }
        return i == s.size();
    }
};

int main(int argc, char const *argv[]) {
    Solution392 solu;
    cout << solu.isSubsequence("abc", "ahbgdc") << endl;
    return 0;
}